crypto/bp/bp_gt.c

```diff
@@ -0,0 +1,313 @@
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ====================================================================
+ *
+ * This product includes cryptographic software written by Eric Young
+ * (eay@cryptsoft.com).  This product includes software written by Tim
+ * Hudson (tjh@cryptsoft.com).
+ *
+ */
+/*
+ * ====================================================================
+ * Copyright 2016 MIRACL UK Ltd., All Rights Reserved. Portions of the
+ * attached software ("Contribution") are developed by MIRACL UK LTD., and
+ * are contributed to the OpenSSL project. The Contribution is licensed
+ * pursuant to the OpenSSL open source license provided above.
+ */
+
+#include <openssl/ec.h>
+
+#include "bp_lcl.h"
+
+GT_ELEM *GT_ELEM_new(const BP_GROUP *group)
+{
+    GT_ELEM *ret;
+
+    if ((ret = OPENSSL_malloc(sizeof(*ret))) == NULL)
+        return NULL;
+
+    ret->f = FP12_new();
+    if (ret->f == NULL) {
+        FP12_free(ret->f);
+        return NULL;
+    }
+    return (ret);
+}
+
+void GT_ELEM_free(GT_ELEM *a)
+{
+    if (a == NULL)
+        return;
+    else
+        FP12_free(a->f);
+    OPENSSL_free(a);
+}
+
+void GT_clear_free(GT_ELEM *a)
+{
+    if (a == NULL)
+        return;
+    else
+        FP12_clear_free(a->f);
+    OPENSSL_free(a);
+}
+
+int GT_ELEM_copy(GT_ELEM *a, const GT_ELEM *b)
+{
+    return FP12_copy(a->f, b->f);
+}
+
+GT_ELEM *GT_ELEM_dup(const GT_ELEM *a, const BP_GROUP *group)
+{
+    GT_ELEM *t;
+    int r;
+
+    if (a == NULL)
+        return NULL;
+    t = GT_ELEM_new(group);
+    if (t == NULL)
+        return NULL;
+    r = GT_ELEM_copy(t, a);
+    if (!r) {
+        GT_ELEM_free(t);
+        return NULL;
+    }
+    return t;
+}
+
+int GT_ELEM_zero(GT_ELEM *a)
+{
+    return FP12_zero(a->f);
+}
+
+int GT_ELEM_is_zero(GT_ELEM *a)
+{
+    return FP12_is_zero(a->f);
+}
+
+int GT_ELEM_set_to_unity(const BP_GROUP *group, GT_ELEM *a)
+{
+    if (!FP12_zero(a->f))
+        return 0;
+    return BN_copy(a->f->f[0]->f[0]->f[0], group->one) != NULL;
+}
+
+int GT_ELEM_is_unity(const BP_GROUP *group, const GT_ELEM *a)
+{
+    if (FP6_is_zero(a->f->f[1]) == 0)
+        return 0;
+    if (FP2_is_zero(a->f->f[0]->f[1]) == 0)
+        return 0;
+    if (FP2_is_zero(a->f->f[0]->f[2]) == 0)
+        return 0;
+    if (BN_is_zero(a->f->f[0]->f[0]->f[1]) == 0)
+        return 0;
+    if (!BN_cmp(a->f->f[0]->f[0]->f[0], group->one) == 0)
+        return 0;
+    return 1;
+}
+
+size_t GT_ELEM_elem2oct(const BP_GROUP *group, const GT_ELEM *a,
+                        unsigned char *buf, size_t len, BN_CTX *ctx)
+{
+    size_t ret;
+    BN_CTX *new_ctx = NULL;
+    int used_ctx = 0;
+    BIGNUM *f;
+    size_t field_len, i, j, k, l, skip;
+
+    /*
+     * ret := required output buffer length
+     */
+    field_len = BN_num_bytes(group->field);
+    ret = 12 * field_len;
+
+    /*
+     * if 'buf' is NULL, just return required length
+     */
+    if (buf != NULL) {
```

Comment:
This can be simplified to get rid of `used_ctx` etc:
```
size_t required_len;
...
required_len = 12 * field_len;
if (buf == NULL)
    return required_len;
/* Now the code inside buf != NULL. */
```